packages/core/src/compiler/Style.ts

```diff
@@ -1370,38 +1171,457 @@ const matchDeclLine = (
+  subProg: SubProg<A>
+): [im.Set<string>, im.List<im.List<[Subst, SubStmt<A> | undefined]>>] => {
+  const initUsedStyVars: im.Set<string> = im.Set();
+  const initListRSubsts: im.List<
+    im.List<[Subst, SubStmt<A> | undefined]>
+  > = im.List();
+
+  const [newUsedStyVars, newListRSubsts] = rels.reduce(
+    ([usedStyVars, listRSubsts], rel) => {
+      const [relUsedStyVars, relRSubsts] = matchStyRelToSubRels(
+        styTypeMap,
+        subTypeMap,
+        varEnv,
+        subEnv,
+        rel,
+        subProg
+      );
+      return [usedStyVars.union(relUsedStyVars), listRSubsts.push(relRSubsts)];
+    },
+    [initUsedStyVars, initListRSubsts]
+  );
+
+  return [newUsedStyVars, newListRSubsts];
+};
+
+/**
+ * First match the relations. Then, match free Style variables. Finally, merge all substitutions together.
+ */
+const makePotentialSubsts = (
+  varEnv: Env,
+  selEnv: SelEnv,
+  subEnv: SubstanceEnv,
   subProg: SubProg<A>,
   decls: DeclPattern<A>[],
-  initSubsts: Subst[]
-): Subst[] => {
-  return decls.reduce(
-    (substs, decl) => matchDecl(varEnv, subProg, substs, decl),
-    initSubsts
+  rels: RelationPattern<A>[]
+): im.List<[Subst, im.Set<SubStmt<A> | undefined>]> => {
+  const subTypeMap: { [k: string]: TypeConsApp<A> } = subProg.statements.reduce(
+    (result, statement) => {
+      if (statement.tag === "Decl") {
+        result[statement.name.value] = statement.type;
+        return result;
+      } else {
+        return result;
+      }
+    },
+    {}
   );
+  const styTypeMap: { [k: string]: StyT<A> } = selEnv.sTypeVarMap;
+  const [usedStyVars, listRSubsts] = makeListRSubstsForStyleRels(
+    styTypeMap,
+    subTypeMap,
+    varEnv,
+    subEnv,
+    rels,
+    subProg
+  );
+  // Add in variables that are not present in the relations.
+  const listPSubsts = decls.reduce((currListPSubsts, decl) => {
+    if (usedStyVars.includes(decl.id.contents.value)) {
+      return currListPSubsts;
+    } else {
+      const pSubsts = matchDecl(varEnv, subProg, decl);
+      return currListPSubsts.push(pSubsts.map((pSubst) => [pSubst, undefined]));
+    }
+  }, listRSubsts);
+
+  if (listPSubsts.some((pSubsts) => pSubsts.size === 0)) {
+    return im.List();
+  }
+  // Note: listPSubsts has type List<List<[Subst, SubStmt]>>
+  // where one Substitution directly refers to one Substance statement.
+
+  // Below, we merge the lists together. A merged substitution may
+  // represent / satisfy multiple Substance statements. Therefore we
+  // convert SubStmt into Set<SubStmt>, and perform the merging.
+
+  const first = listPSubsts.first();
+  if (first) {
+    // first_ converts List<[Subst, SubStmt]> into List<[Subst, Set<SubStmt>]>
+    const first_: im.List<
+      [Subst, im.Set<SubStmt<A> | undefined>]
+    > = first.map(([subst, stmt]) => [
+      subst,
+      im.Set<SubStmt<A> | undefined>().add(stmt),
+    ]);
+    const substs = listPSubsts.shift().reduce((currSubsts, pSubsts) => {
```

Comment:
I know `shift` is linear-time for JS arrays; is it the same for `im.List`, or is it faster (constant or logarithmic time)?

Reply:
I'm not sure whether or not it's linear time. But it shouldn't have a big effect, given that this will be run only once for each block, and the list it operates on has size O(num of Stu Relations + num of Sty Variables).

Reply:
Makes sense!